xdsl/dialects/builtin.py

```diff
@@ -731,27 +734,27 @@ def verify(self):
 
     @staticmethod
     def create_dense_int_or_index(typ: IntegerType | IndexType,
-                                  data: List[int | IntAttr]) -> DenseArrayBase:
-        attr_list = [IntAttr(d) if isinstance(d, int) else d for d in data]
+                                  data: Sequence[int]) -> DenseArrayBase:
+        attr_list = [IntAttr(d) for d in data]
         return DenseArrayBase([typ, ArrayAttr(attr_list)])
 
     @staticmethod
-    def create_dense_float(
-            typ: Float16Type | Float32Type | Float64Type,
-            data: List[int | float | FloatData]) -> DenseArrayBase:
-        data_attr = [
-            FloatData(float(d)) if isinstance(d, float | int) else d
-            for d in data
-        ]
+    def create_dense_float(typ: Float16Type | Float32Type | Float64Type,
+                           data: Sequence[int | float]) -> DenseArrayBase:
+        data_attr = [FloatData(float(d)) for d in data]
         return DenseArrayBase([typ, ArrayAttr(data_attr)])
 
     @staticmethod
-    def from_list(
-        type: Attribute, data: List[int | AnyIntegerAttr]
-        | List[int | float | AnyFloatAttr]
-    ) -> DenseArrayBase:
+    def from_list(type: Attribute,
+                  data: Sequence[int | float]) -> DenseArrayBase:
         if isinstance(type, IndexType | IntegerType):
-            return DenseArrayBase.create_dense_int_or_index(type, data)
+
+            def type_map(el: int | float) -> int:
+                assert isinstance(el, int)
+                return el
+
+            return DenseArrayBase.create_dense_int_or_index(
+                type, [type_map(d) for d in data])
```

Comment:
nit: Can we maybe make this a little bit nicer with `@overload`?

Reply:
I agree, though I feel that this is a broader discussion than this PR. In general, DenseArrayBase, and DenseIntOrFloatElementsAttr before it, both have the issue of not quite being clear, at least in our codebase. You cannot quite infer just from the argument types what the user wants to get, unless you use a lot of constructors, where the name of the constructor does that job, or you let the user pass in some other argument.

I would be interested in that discussion in general though, but maybe in a later PR :)

Reply:
Yeah, I'd say this can be continually cleaned up. It's quite a complex class, with the way the different fields and constructors interact with each other. Hopefully these changes make for a small simplification, we can then see how to improve it further.